### DISCUSSION TITLE
How to use Resolving random private addresses with IRK ?

### DISCUSSION THREAD

**Discussion by arnog23** [theengs/gateway#207] [Q&A]
Hello,

How to use this feature to detect iOS devices if possible ?

How to find the IRK ? Do we have to manually choose it ? Which MAC should we use ? The private MAC available in General=> Settings ?

Thanks

**Comment by DigiH** [comment]
Hi @arnog23 

This feature so far is only fully available with the development branch of Theengs Gateway, but will be in the next upcoming release.

The update will also include the updated documentation, including instructions on how to get the Identity MAC Address and IRK, which you can already view here

https://github.com/theengs/gateway/blob/development/docs/use/use.md#getting-identity-resolving-key-irk-for-apple-watch-iphone-and-ipad

Let us know if you think it is all clear, or if you might have any improvement instructions.

**Reply by arnog23** [reply]
Hello @DigiH 
Yes it seems to be clear but it also seems to be clear that it won't be easy without MacOS 😁 except using third apps but we need to be sure they are safe.

**Reply by DigiH** [reply]
Yes, currently the instructions require a MAC to be able to access the IRK, which is what I used here to retrieve them to implement and test the Apple decoders, but I think there might be other ways to retrieve them which I have to investigate more.

I might need to do a few small adjustments to the iPhone/iPad decoder before we're happy to issue a new Theengs Gateway release, but I should be done with testing this week.

Regarding safety, with the macOS keychain requiring your password twice and hoping that the Apple Watch, iPhone and iPad are also set with passwords, retrieving this information should be secured. The same I assume for the device tghen running Theengs Gateway with this information entered.

May I ask which of the Apple devices you have, and don't you have a Mac?

**Reply by arnog23** [reply]
I do not have a mac but I have iPhone, iPad and iWatch.

**Reply by DigiH** [reply]
Have a look at this HA forum posting, which requires an ESP32 to retrieve the IRKs without a Mac.

https://community.home-assistant.io/t/guide-getting-the-irk-for-iphone-and-apple-watch-without-macos-maybe-also-other-devices/623694

I have not tried it myself, so let us know how you get on. Once you have the IRKs for the Identity MAC addresses of your devices just note them down to use with the upcoming Theengs Gateway release.

I can't remember how you have Theengs Gateway installed for use with Jeedom from our Oral-B toothbrush conversation, but if you have a pip install we have also made it easier to download and install the latest development versions of Theengs Gateway and Theengs Decoder, without having to build them yourself.

**Reply by arnog23** [reply]
Interesting for the ESP32 but I do not have one and it also needs a third app (is it a secure app ?).
For the gateway, I do not have the testing one anymore but I can create a new one if needed.

**Reply by DigiH** [reply]
> is it a secure app ?

I would hope so, but cannot vouch for it, as I have not tested it myself to see if it has any suspicious outside communication.

An easy alternative could also be, if you have friends with a Mac, to ask them to temporarily let you set up your own user account on their Mac, with your APple ID, so you can retrieve the IRKs as per our macOS Keychain instructions.

Then afterwards delete the whole user account on their Mac to completely delete any traces.

**Reply by DigiH** [reply]
BTW, what PC OS do you have, running Linux or Windows?

**Reply by arnog23** [reply]
I don't know how this app can be noted with 5 in the store with only 4 comments in 4 year and not good ones ...

Both, Linux and Windows.

**Reply by DigiH** [reply]
On Windows you could also get the iCloud entries form the iCloud Keychain, on Windows called iCloud Passwords AFAIK.

As the iCloud syncing is encrypted you should not worry about security, and once the IRKs retrieved you could also delete the app again from your Windows system.

I would assume that Password and Keychain syncing would have to be enabled on your iPad and iPhone in the iCloud settings, again something you could turn off again afterwards. Or possibly through locally syncing your iPhone and iPad with iTunes/Music on Windows.

A good article about this seems to be

https://appleinsider.com/inside/icloud/tips/how-to-use-icloud-keychain-on-windows-and-how-it-differs-from-macos-and-ios

If you are successful with this we could add additional Windows instructions to the Gateway documentation with your help.

**Reply by arnog23** [reply]
iCloud Password on Windows needs Windows Hello to work. Unfortunately, I do not see any iOS device account  into it. Only websites credentials.

**Reply by DigiH** [reply]
Hmm, did it ask you to log in with your Apple ID credentials, and are you logged into iCloud on your devices?

I have to admit I have absolutely no experience with iCloud Password on Windows or Windows Hello.

It really looks like you need to opt for any of the other possibilities mentioned above.

**Reply by arnog23** [reply]
Yes, to be able to access credentials from iCloud password app, you need to be logged-in with iCloud account into the iCloud Windows app with passwords option enabled.
